ipam/pool_test.go

```diff
@@ -0,0 +1,991 @@
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					ifName := "eth0"
+					as.Pools[poolId] = &addressPool{
+						Id:     poolId,
+						IfName: ifName,
+					}
+					options := map[string]string{}
+					options[OptInterfaceName] = "en0"
+					ap, err := as.requestPool("", "", options, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When as has one pool available", func() {
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "poolId"
+					as.Pools[poolId] = &addressPool{
+						Id: poolId,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When as has pools with different priority", func() {
+				It("Should return the pool with the highest priority", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["pool1"] = &addressPool{
+						Id:       "pool1",
+						Priority: 1,
+					}
+					as.Pools["pool2"] = &addressPool{
+						Id:       "pool2",
+						Priority: 2,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("pool2"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When as has pools with different addresses", func() {
+				It("Should return the pool with the highest number of addresses", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["pool1"] = &addressPool{
+						Id:        "pool1",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["pool2"] = &addressPool{
+						Id:        "pool2",
+						Addresses: map[string]*addressRecord{},
+					}
+					as.Pools["pool2"].Addresses["ar"] = &addressRecord{}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("pool2"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+		})
+
+		Describe("Test releasePool", func() {
+			Context("When pool not found", func() {
+				It("Should return error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					err := as.releasePool("pool")
+					Expect(err).To(Equal(errAddressPoolNotFound))
+				})
+			})
+
+			Context("When poll is not in use", func() {
+				It("Should return error", func() {
+					poolId := "pool"
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools[poolId] = &addressPool{
+						RefCount: 0,
+					}
+					err := as.releasePool("pool")
+					Expect(err).To(Equal(errAddressPoolNotInUse))
+				})
+			})
+
+			Context("", func() {
+				It("", func() {
```

Comment:
We may forget the expectation.

Reply:
thx, it should be "When pool's epoch is less than the space's epoch and pool is never in use"